src/tools/miri/src/concurrency/sync.rs

```diff
@@ -128,6 +128,8 @@ struct FutexWaiter {
     thread: ThreadId,
     /// The bitset used by FUTEX_*_BITSET, or u32::MAX for other operations.
     bitset: u32,
+    /// Extra info stored for this waiter.
+    extra: Option<u32>,
```

Comment:
So this field encodes whether the waiter is PI waiter? Just calling it `extra` doesn't make that very clear.

Reply:
> We need to store tid here because we need to write the tid of the top waiter to futex when waking it.

Why can't the thread-that-is-woken-up do that itself in the wakeup callback?

Reply:
> Why can't the thread-that-is-woken-up do that itself in the wakeup callback?

It's possible; however it requires some tight coupling logic in `this.futex_wait()` that checks whether we are a PI waiter, grabs the tid, and writes the futex. I'd prefer the current implementation.

Reply:
I'll try to keep this in mind for the review of the PI paths, to see if I can find an elegant alternative.